src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,391 @@
+            }
+
+            [StructLayout(LayoutKind.Sequential, Pack=1)]
+            private unsafe struct AndroidTzDataEntry
+            {
+                public fixed byte id[40];
+                public int byteOffset;
+                public int length;
+                public int unused; // Was raw GMT offset; always 0 since tzdata2014f (L).
+            }
+
+            private string[] _ids;
+            private int[] _byteOffsets;
+            private int[] _lengths;
+
+            public AndroidTzData()
+            {
+                string tzFilePath = GetTimeZoneDirectory() + TimeZoneFileName;
+                using (FileStream fs = File.OpenRead(tzFilePath))
+                {
+                    ReadHeader(tzFilePath, fs);
+                }
+            }
+
+            [MemberNotNull(nameof(_ids))]
+            [MemberNotNull(nameof(_byteOffsets))]
+            [MemberNotNull(nameof(_lengths))]
+            private unsafe void ReadHeader(string tzFilePath, Stream fs)
+            {
+                int size = Math.Max(Marshal.SizeOf(typeof(AndroidTzDataHeader)), Marshal.SizeOf(typeof(AndroidTzDataEntry)));
+                Span<byte> buffer = stackalloc byte[size];
+                AndroidTzDataHeader header = ReadAt<AndroidTzDataHeader>(tzFilePath, fs, 0, buffer);
+
+                header.indexOffset = NetworkToHostOrder(header.indexOffset);
+                header.dataOffset = NetworkToHostOrder(header.dataOffset);
+
+                // tzdata files are expected to start with the form of "tzdata2012f\0" depending on the year of the tzdata used which is 2012 in this example
+                if (header.signature[0] != (byte)'t' || header.signature[1] != (byte)'z' || header.signature[2] != (byte)'d' || header.signature[3] != (byte)'a' || header.signature[4] != (byte)'t' || header.signature[5] != (byte)'a' || header.signature[11] != 0)
+                {
+                    var b = new StringBuilder();
+                    for (int i = 0; i < 12; ++i) {
+                        b.Append(' ').Append((header.signature[i]).ToString("x2"));
+                    }
+
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_BadTZHeader, tzFilePath, b.ToString()));
+                }
+
+                ReadIndex(tzFilePath, fs, header.indexOffset, header.dataOffset, buffer);
+            }
+
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2087:UnrecognizedReflectionPattern",
+                Justification = "Implementation detail of Android TimeZone")]
+            private unsafe T ReadAt<T>(string tzFilePath, Stream fs, long position, Span<byte> buffer)
+                where T : struct
+            {
+                int size = Marshal.SizeOf(typeof(T));
+                Debug.Assert(buffer.Length >= size);
+
+                fs.Position = position;
+                int numBytesRead;
+                if ((numBytesRead = fs.Read(buffer)) < size)
```

Comment:
This should be a while loop. The contract for `Read` allows it to read less then requested. (From https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.read?view=net-5.0: An implementation is free to return fewer bytes than requested even if the end of the stream has not been reached.)

Reply:
What would this need to loop over? It seems like filling up the buffer is sufficient for grabbing the bytes corresponding to either the timezone data header or the timezone data entry which can be accomplished with one `Read` unless the end of the buffer is hit.

Reply:
It should loop until the required number of bytes is read; or until Read returns 0.

Reply:
We added a loop to read until the number of bytes expected in either the tz data header or tz data entry has been reached. `fs.Read()` returns 0 when the end of tzdata has been reached, which should only occur when we read the last entry in `tzdata`. 

The parsing iteratively reads the file as header, entry, entry, entry ...
And with recent changes, is done through `LoadTzFile` and `ReadIndex`.